### PR TITLE
feat(stream): first version of mysql cdc connector

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Add the CDC connector source, which can ingest data from a MySQL database.
The CDC connector follows our framework of data source, which consists of a `CdcSplitEnumerator`, `CdcSplit` and `CdcSplitReader`. The `CdcSplitReader` is responsible for communicating with the Connector Node to create the CDC engine and pull the data change events.

**Limitations**
- CDC source only supports a single split now, so only one source executor can consume events
- CDC source doesn't support failover now (relying on the CDC engine to support rewinding its offset)

Example SQL for creating a CDC source:
```
dev=> create materialized source orders (
   order_id int,
   order_date bigint,
   customer_name string,
   price decimal,
   product_id int,
   order_status smallint,
   PRIMARY KEY (order_id)
) with (
 connector = 'cdc',
 database.hostname = '127.0.0.1',
 database.port = '3306',
 database.user = 'root',
 database.password = '123456',
 database.name = 'mydb',
 table.name = 'orders'
) row format debezium_json;
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/connector/src/source/cdc/source/reader.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+use risingwave_common::util::addr::HostAddr;
+use risingwave_pb::cdc_service::{DbConnectorProperties, GetEventStreamResponse};
+use risingwave_rpc_client::CdcClient;
+
+use crate::source::base::{SourceMessage, SplitReader};
+use crate::source::cdc::{CdcProperties, CdcSplit};
+use crate::source::{BoxSourceStream, Column, ConnectorState, SplitImpl};
+
+// FIXME: put the connector node addr to config file
+const CDC_NODE_ENDPOINT: &str = "127.0.0.1:60061";
```
Comment (liurenjie1024) [inline src/connector/src/source/cdc/source/reader.rs:31]: Please make this configurable, at least the port should be configurable.

File: src/connector/src/source/cdc/enumerator/mod.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+use async_trait::async_trait;
+
+use crate::source::cdc::{CdcProperties, CdcSplit};
+use crate::source::SplitEnumerator;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct CdcSplitEnumerator {
+    source_id: u32,
+    split_num: i32,
```
Comment (liurenjie1024) [inline src/connector/src/source/cdc/enumerator/mod.rs:23]: If currently `split_num` is not used, how about remove it?

File: src/connector/src/source/cdc/source/reader.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+// FIXME: currently only support single split for CDC source
+pub struct CdcSplitReader {
+    assigned_splits: HashMap<String, Vec<CdcSplit>>,
```
Comment (liurenjie1024) [inline src/connector/src/source/cdc/source/reader.rs:35]: Remove this unused field?

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM

File: src/connector/src/source/cdc/source/reader.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                    database_name: props.database_name.clone(),
+                    table_name: props.table_name.clone(),
+                    partition: props.parititon.clone(),
+                    start_offset: props.start_offset.clone(),
+                    include_schema_events: false,
+                },
+            )
+            .await?;
+        pin_mut!(cdc_stream);
+        while let Some(event_res) = cdc_stream.next().await {
```
Comment (tabVersion) [inline src/connector/src/source/cdc/source/reader.rs:88]: for_await

File: src/connector/src/source/cdc/source/reader.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    async fn new(
+        props: CdcProperties,
+        state: ConnectorState,
+        _columns: Option<Vec<Column>>,
+    ) -> Result<Self> {
+        if let Some(splits) = state {
+            // expect only one split
+            debug_assert!(splits.len() == 1);
+            for split in &splits {
+                if let SplitImpl::Cdc(cdc_split) = split {
```
Comment (tabVersion) [inline src/connector/src/source/cdc/source/reader.rs:47]:
```suggestion
        let cdc_split = state.exactly_one().map_err(...)?;
```

File: src/connector/src/source/cdc/mod.rs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#[derive(Clone, Debug, Deserialize)]
+pub struct CdcProperties {
+    #[serde(default)]
+    pub connector_node_addr: String,
+    #[serde(default)]
+    pub source_id: u32,
+    #[serde(default)]
+    pub start_offset: String,
+    #[serde(default)]
+    pub parititon: String,
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/connector/src/source/cdc/mod.rs:35]: Messages produced by debezium contain the partition field. It is a reserved field now, may be removed in future.

File: src/connector/src/source/cdc/mod.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    #[serde(rename = "table.name")]
+    pub table_name: String,
+    #[serde(rename = "database.hostname")]
+    pub database_host: String,
+    #[serde(rename = "database.port")]
+    pub database_port: String,
+    #[serde(rename = "database.user")]
+    pub database_user: String,
+    #[serde(rename = "database.password")]
+    pub database_password: String,
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/connector/src/source/cdc/mod.rs:47]: These properties come directly from the WITH clause in SQL. And the `impl_connector_properties` macro relies on `serde_json` to parse those `key: value` pairs to the corresponding connector properties. If replaced with a hashmap here, I think the macro cannot work.

File: src/connector/src/source/cdc/source/reader.rs (hunk 5 above)
Comment (StrikeW) [reply src/connector/src/source/cdc/source/reader.rs:47]: Sorry, I do not find this `exactly_one` function in `state` or `splits`. something wrong?

**Review by liurenjie1024 — APPROVED** [review]
LGTM

File: src/connector/src/source/cdc/source/reader.rs (hunk 5 above)
Comment (tabVersion) [reply src/connector/src/source/cdc/source/reader.rs:47]: you may need `use itertools` for getting the element out of the vec
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.exactly_one

File: src/connector/src/source/cdc/source/message.rs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+use risingwave_pb::cdc_service::CdcMessage;
+
+use crate::source::base::SourceMessage;
+
+impl From<CdcMessage> for SourceMessage {
+    fn from(message: CdcMessage) -> Self {
+        SourceMessage {
+            payload: Some(Bytes::from(message.payload)),
+            offset: message.offset,
+            split_id: message.partition.into(),
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/connector/src/source/cdc/source/message.rs:25]: As I mentioned above, debezium produces the partition field for each event, here is a combination of `database_name.table_name` for MySQL. We may have other usages for it.

File: src/connector/src/source/cdc/source/reader.rs (hunk 5 above)
Comment (StrikeW) [reply src/connector/src/source/cdc/source/reader.rs:47]: Got it. Thanks

**Pull request merged**